### PR TITLE
Add RelatedFactoryList class for one-to-many support

### PR DESCRIPTION
**Pull request by seanharr11**
The ``RelatedFactoryList`` class allows us to define a declaration which
returns a list of RelatedFactory(s).

This is different than the current RelatedFactory, which only generates
1 related object when it's parent Factory is 'created'.

This is useful for one-to-many relations. For instance, let's pretend we
have a ``Student`` model which has *many* ``Exams``. We don't want to
strictly generate *one* exam per student in our test suite, we want
to cover the case of many exams per student.

So, we'd tell the ``Student.exams`` declaration to be a
``RelatedFactoryList``, and could either tell it to generate ``length``
exams per student, or provide a ``lambda`` which would lazily get
evaluated; this allows for a dynamic number of related objects
being generated for each parent object.

This lambda can be used to create deterministic counts of related
objects per parent object, or to randomly generate "n" related
objects, where "n" is a random int. (See added test coverage
for the former case, and latter case).

This has been brought up in a few StackOverflow posts. I wasn't
satisfied so I opened this PR!

http://stackoverflow.com/questions/27120685/one-to-many-relationships-in-factory-boy/
http://stackoverflow.com/questions/14589362/creating-multiple-objects-with-foreign-key/

### CODE REVIEW

**Comment by seanharr11** [comment]
By the way, I named the branch 'related-field-list' instead of 'related-factory-list'...my fault.

File: factory/declarations.py [hunk 1]
```diff
@@ -624,6 +624,28 @@ def call(self, instance, step, context):
...
+            returning a list of 'factory' objects w/ length 'length'.
+    """
+
+    def __init__(self, factory, factory_related_name='', length=2, **defaults):
+        self.length = length
+        super(RelatedFactoryList, self).__init__(factory, factory_related_name, **defaults)
+
+    def call(self, instance, step, context):
+        return [super(RelatedFactoryList, self).call(instance, step, context)
+                for i in range(0, self.length if isinstance(self.length, int)
```
Comment (jeffwidman) [inline factory/declarations.py]: the `0` is superfluous as it's the default

File: tests/test_using.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
 import os
 import sys
 import warnings
+import random
```
Comment (jeffwidman) [inline tests/test_using.py]: alphabetical order please within a section, ie, this should be above `sys`

File: tests/test_using.py [hunk 3]
```diff
@@ -2009,6 +2010,134 @@ class Meta:
...
+        class TestRelatedObjectFactoryList(factory.Factory):
+            class Meta:
+                model = TestRelatedObject
+            one = 1
+            two = factory.LazyAttribute(lambda o: o.one + 1)
+
+        class TestObjectFactory(factory.Factory):
+            class Meta:
+                model = TestObject
+            one = 3
```
Comment (jeffwidman) [inline tests/test_using.py]: these variable names are confusing when skimming the code... one what? At first glance I assume this is integers, so why is `one` set to `3`?

Is there a slightly-more verbose naming scheme that makes it a little more clear of the intent of these variables?

Comment (seanharr11) [inline tests/test_using.py]: @jeffwidman - I agree with you that these variable names are very confusing. That said, I was just using the naming convention used in other tests in the test_using.py file. If you take a look at the rest of the tests in that file, they all use ``one``, ``two``, etc...I wanted to remain consistent.

Comment (jeffwidman) [inline tests/test_using.py]: >  I was just using the naming convention used in other tests in the test_using.py file. If you take a look at the rest of the tests in that file, they all use one, two, etc...I wanted to remain consistent.

Oh, yeah, consistency makes total sense here.

**Comment by rbarrois** [comment]
Thanks for the suggestion! This looks great indeed.

I have a few additional suggestions / remarks for the API:
- I'd rather use the same parameter for the batch size as in [create_batch](http://factoryboy.readthedocs.io/en/latest/reference.html#factory.Factory.create_batch): `size`
- Instead of requiring either an int or a lambda, what about allowing any factory-style declaration? One could thus use a `FuzzyInteger` or pick the size from a parameter, etc.

And sorry for taking so long to review this :disappointed:

File: docs/reference.rst [hunk 4]
```diff
@@ -1703,6 +1704,58 @@ If a value is passed for the :class:`RelatedFactory` attribute, this disables
...
+    .. attribute:: name
+
+        The generated object (where the :class:`RelatedFactory` attribute will
+        set) may be passed to the related factories if the :attr:`factory_related_name` parameter
+        is set.
+
+        It will be passed as a keyword argument, using the :attr:`name` value as
+        keyword:
+
+    .. attribute:: length
```
Comment (imomaliev) [inline docs/reference.rst]: I suggest using already existing terms, for `create_batch`, `build_batch` etc it is `size`

**Comment by seanharr11** [comment]
@jeffwidman are we able to merge this in? Have an internal project that we're trying to open-source and it'd be nice to have `requirements.txt` reflecting the master branch, rather than my forked branch! Thanks man!

**Pull request closed without merging**